docs/NNNN-hash-and-id-n-tuple-storage-layout.md

```diff
@@ -0,0 +1,380 @@
+of any length, containing any characters, to OCFL object root directories.
+
+Using this extension, OCFL object identifiers are hashed and encoded as hex
+strings (all letters lower-case). These digests are then divided into _N_
+n-tuple segments, which are used to create nested paths under the OCFL storage
+root. Finally, the OCFL object identifier is percent-encoded to create a
+directory name for the OCFL object root (see ["Encapsulation
+Directory"](#encapsulation-directory) section below).
+
+The n-tuple segments approach allows OCFL object identifiers to be evenly
+distributed across the storage hierarchy. The maximum number of files under any
+given directory is controlled by the number of characters in each n-tuple, and
+the tree depth is controlled by the number of n-tuple segments each digest is
+divided into. The encoded encapsulation directory name provides visibility into
+the object identifier from the file path (see ["Encapsulation
+Directory"](#encapsulation-directory) section below for details).
+
+## Encapsulation Directory
+
+For basic OCFL object identifiers, the object identifier is used as the name of
+the encapsulation directory (ie. the object root directory).
+
+Some object identifiers could contain characters that are not safe for directory
+names on all filesystems. Safe characters are defined as A-Z, a-z, 0-9, '-' and
+'\_'. When an unsafe character is encountered in an object identifier, it is
+percent-encoded using the lower-case hex characters of its UTF-8 encoding.
+
+Some object identifiers could also result in an encoded string that is longer
+than can be supported as a directory name. To handle that scenario, if the
+percent-encoded object identifier is longer than 100 characters, it is truncated
+to 100 characters, and then the digest of the original object identifier is
+appended to the encoded object identifier like this:
+<encoded-object-identifier-first-100-chars>-<digest>. Note: this means that it
+is no longer possible to determine the full object identifier from the
+encapsulation directory name - some characters have been removed, and even the
+first 100 characters of the encoded object identifier cannot be fully, reliably
+decoded, because the truncation may leave a partial encoding at the end of the
+100 characters.
+
+| Object ID | Encapsulation Directory Name |
+| --- | --- |
+| object-01 | object-01 |
+| ..Hor/rib:lè-$id | %2e%2eHor%2frib%3al%c3%a8-%24id |
+| abcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghija | abcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghijabcdefghij-5cc73e648fbcff136510e330871180922ddacf193b68fdeff855683a01464220 |
+
+## Parameters
+
+### Summary
+
+* **Name:** `digestAlgorithm`
+  * **Description:** The digest algorithm to apply to the OCFL object identifier; MUST be an algorithm that is allowed in the OCFL fixity block
+  * **Type:** string
+  * **Constaints:** Must not be empty
```

Comment:
typo (found from diff): should be `Constraints`

Reply:
thanks, fixed in b697de6